pkg/chrootarchive/archive_unix.go

```diff
@@ -79,6 +80,11 @@ func invokeUnpack(decompressedArchive io.Reader, dest string, options *archive.T
 	w.Close()
 
 	if err := cmd.Wait(); err != nil {
+		// when `xz -d -c -q | docker-untar ...` failed on docker-untar side,
+		// we need to exhaust `xz`'s output, otherwise the `xz` side will be
+		// pending on write pipe forever
+		io.Copy(ioutil.Discard, decompressedArchive)
```

Comment:
first thing, are you able to reproduce issue #16924 crash ?

then, apply the changes here, but without this line of draining decompressedArchive, then docker daemon is hanging on xz, in pstree it shows xz as a child process of docker daemon, is sleeping state, if `strace -p <xz-pid>`, then you see xz is pending on `write(1, "....", 8192)` , that fd 1 is pipe fd connect to docker daemon (and connect to reexec docker-untar)